File: 1254-number-of-closed-islands/1254-number-of-closed-islands.cpp
class Solution {
private:
 
    int dr[4] = {0, 0, 1, -1};
    int dc[4] = {-1, 1, 0, 0};

    bool boundry;

    bool isValid(int r, int c, int ROWS ,int COLS){
        if(r < 0 || r >= ROWS) return false;
        if(c < 0 || c >= COLS) return false;
        return true;
    }

    bool atBoundries(int r , int c , int ROWS , int COLS){
        if(r == 0 || c == 0 || r  == ROWS-1 || c == COLS-1) return true;
        return false;
    }

    void dfs(int r , int c , vector<vector<int>>& grid){
        // out of boundries
        if(!isValid(r,c,grid.size() , grid[0].size()) || grid[r][c] == 1 ) return ;
        // touches water
        if(atBoundries(r,c,grid.size(),grid[0].size())){
            boundry = true;
        }
        // mark it as visited
        grid[r][c] = 1;
        for(int d = 0 ; d < 4 ; d++){
            dfs(r + dr[d] , c +dc[d] , grid) ;
        }
        
    }

public:
    int closedIsland(vector<vector<int>>& grid) {


        int ROWS = grid.size(), COLS = grid[0].size();
        int numOfClosedIslands {0};
        for(int r = 0 ;  r < ROWS ; r++){

            for(int c = 0 ; c < COLS ; c++){
                if(grid[r][c] == 0){
                    boundry = false;
                    dfs(r,c,grid);
                    numOfClosedIslands += !boundry;
                    
                }
                
            }
        }
        return numOfClosedIslands;

    }
};

File: 1254-number-of-closed-islands/1254-number-of-closed-islands.py
class Solution:
    def closedIsland(self, grid: List[List[int]]) -> int:

        def isValid(r, c, ROWS, COLS):
            if r < 0 or r >= ROWS: return False
            if c < 0 or c >= COLS: return False
            return True

        def atBoundries(r, c, ROWS, COLS):
            if r == 0 or c == 0 or r == ROWS-1 or c == COLS-1: return True
            return False

        def dfs(r, c, grid):
            if not isValid(r, c, len(grid), len(grid[0])) or grid[r][c] == 1: return
            if atBoundries(r, c, len(grid), len(grid[0])):
                self.isAtBoundry = True
            grid[r][c] = 1
            for i in range(4):
                dfs(r + dr[i], c + dc[i], grid)

        ROWS, COLS = len(grid), len(grid[0])
        numOfClosedIslands = 0
        dr = [0, 0, 1, -1]
        dc = [-1, 1, 0, 0]
        for r in range(ROWS):
            for c in range(COLS):
                if not grid[r][c]:
                    self.isAtBoundry = False
                    dfs(r, c, grid)
                    numOfClosedIslands += not self.isAtBoundry
        return numOfClosedIslands



